units.md

```diff
@@ -0,0 +1,83 @@
+
+Gaz is a virtual unit used to measure the computational complexity of executing transactions on the Massa blockchain. Each operation requires a certain amount of Gaz, and the amount of Gaz consumed is proportional to the computational complexity of the operation. In order to make this unit more tangible for the end user, gaz is measured in liters (L).
+Gaz is expressed as an unsigned 64-bit integer.
+
+1 Liter (L) is equivalent to 1,000,000,000 nanoliter, the smallest unit of gaz.
+
+To simplify usage, the following standard units implemenation are recommended:
+
+- milliliter (mL): 1 milliliter represents 1,000,000 nanoliter or 1/1000th of a liter.
+- microliter (µL): 1 microliter represents 1,000 nanoliter or 1/1,000,000th of a liter.
+- nanoliter (nL): 1 nanoliter represents the smallest unit of gaz or 1/1,000,000,000th of a liter.
+
+### Roll
+
+Roll is a specific token used for staking on the Massa blockchain. A Roll token is required to create and endorse a block on the blockchain. The probability of being drawn to create or endorse a block is correlated to the number of rolls possessed.
+Roll is expressed as an unsigned 64-bit integer.
+
+To simplify usage, the following standard units are recommended:
+
+- kiloRoll (kROLL): 1 kiloRoll represents 1,000 rolls.
+- megaRoll (MROLL): 1 megaRoll represents 1,000 kiloRolls or 1,000,000 rolls.
+
+### Transaction Fee
+
+A transaction fee is a small amount of Massa coin paid by the sender of a transaction to compensate the nodes in the Massa network for validating and executing the transaction.
+
+Transaction fees are denominated in Massa coin and are paid by the sender in addition to the amount being transacted. The fee amount is determined by the sender and is included in the transaction data.
+
+Note: Developers should use the same units implementation as for Massa coin values when handling transaction fee values.
+
+## Usage
+
+When interacting with the Massa blockchain and its objects, it is recommended to use the native types provided by the programming language when available. For instance, when working with the Massa coin, an unsigned 64-bit integer can be used to represent the coin's value.
+
+To make code more readable and avoid errors, it is recommended to use useful constants such as mMASSA or kROLL to express values. For instance, to represent 1 milliMassa, one can write:
+
+```
+const amount = 1 * mMASSA;
+```
+
+In JavaScript, which does not natively support unsigned 64-bit integers, the BigInt type should be used to represent Massa objects such as the Massa coin or Roll tokens. For example, to represent 1 Massa coin, one can write:
+
+```javascript
+const amount = 1n * mMASSA;
+```
+
+In programming languages that allow underscores to be used in numbers for better readability, such as TypeScript and Rust, it is recommended to use them when expressing Massa values. For example, to represent 12,500 milliMassa, one can write:
+
+```typescript
+const amount = 12_500n * mMASSA;
```

Comment:
To me, but this is only a personal preference, this whole notation is super confusing for a user. It requires users to think about applying correct multiplicators in order to scale their amount correctly. I believe this whole idea is inspired by the golang where e.g. time units are implicitely multiply-able. In web3 libraries, in general, I have hardly seen this way of expressing metrics. 
Also,  I believe that users might easily forget to use these constants and just decide / or feel forced to use a whole number BigInt just for their convenience to not be bothered to do maths.
Multiple chains have solved that issue differently. One way to do so is what we already did in web3 by providing the `toMAS` method that takes any BigInt, BigNum, string, number etc. and converts it to the smallest unit (nano). This way users could easily just write fromMAS("0.350") and this would make sure they are sending 0.350 massa in the proper format along with their transaction. Many chains incl. Solana and NEAR do that.

What could have worked even better (again in my opinion only) and what I originally had in web3 is having a strictly typed massa-unit storage class and a converter at the same time. Due to its typed-checked nature, users would be bound to select both amount and units when creating their massa-unit. E.g. `const coins = MassaAmount("1.350", MassaUnit.Massa)` or `const fee = MassaAmount("0.04", MassaUnit.Massa)`. This way inside the code, no matter what the user input is, we could easily convert it in whatever unit we need to send along the wire. It also bounds users to stick to the MassaAmount as the only way to receive and send units through the interface minimizing the sources of errors. Using that way, if necessary, all the metric systems suggested below could be easily integrated be extending the units enum. The MassaAmount would also offer a `toConverter(MassaUnit)` method to facilitate easy conversion in between the units for the enduser.

Reply:
@Thykof good points. Yes. What about the class approach ?

Reply:
I understand the benefit of the strictly typed massa-unit but in my opinion it is not worth it because of the main drawback: the code developers need to write is more verbose everywhere.

but anyway we can't talk about class here in the standard because class is more the implementation part, the standard have to be implemented even in a language that it no Object Oriented

Reply:
This approach is guided by the readability of the code:

```typescript
const coins = MassaAmount("1.350", MassaUnit.Massa)
#vs
const coins = 1_350n * milliMassa
```

My approach is:
- shorter
- doesn't mix decimal with unsign integer
- transparent (no need for the dev to understand what is behind the function MassaAmount or it's returned values)

Reply:
@Thykof Even if the language is not OO, the idea of MassaAmount is still to provide convenience for submitting hard to express units in another unit. Instead of a class, in every language, one could still construct a method called e.g. convertMassaAmount("1.35", MassaUnit.Massa). I dont hardly insist on the MassaAmount class (the idea of it was to wrap extra conversion functionalities), but I would assume there will be a need to at least give users alongside with Greg's idea (1_350n * milliMassa) further methods to perform conversions or easily express their units. Now we have in  massa-web3 `fromMAS` and `toMAS` doing exactly that for coins to nano and vice-versa. I reckon we could provider a wider palette of util conversion methods for such purposes alongside with the alternative notation with the constants - this is just an idea, not a must or anything. For now, I am resolving this discussion as finalized.